Update ABL scripts to use CORGI

TylerZeroMaster opened the pull request:
part of openstax/ce#2182

This new script effectively performs the same steps as the old transformation script:
- Fetch ABL from https://corgi.ce.openstax.org/api/abl/ 
   - Fetches entries intended for REX where code_version <= `REACT_APP_ARCHIVE`
- Select the newest version of each book
- Select versions that are not in config.books.json

`REACT_APP_ARCHIVE` is read from `src/config.archive-url.json`

The tests show the input and output formats (the output format is the same as before: `{ <uuid>: <commit_sha>, ... }`)

TylerZeroMaster committed:
Add script to parse CORGI ABL; update import tests

- Remove old test data
- Add new test data
- Update snapshots with new script

NOTE: The multi-version test data does not include multiple code versions
because the CORGI endpoint will support a query argument that filters by
code version (returns <= code_version)

TylerZeroMaster committed:
Update update-content script to use new transformation

TylerZeroMaster committed:
Improve test data for validate-abl-import

Add additional commit for each book in multi-version test

When there is more than one commit for a book, the jq filter should select
the newest commit

In the test data, the newest commit for each book is bbbbbbb. The books in the
snapshot should all use this commit.

TylerZeroMaster committed:
Show error if fetching ABL fails

TomWoodward commented on script/validate-abl-import/test.bash:
```diff
@@ -27,8 +27,10 @@ for input_file in "$my_dir"/*.input.json; do
     # -------------------------
     json=$(cat "$input_file")
-    node "$my_dir/../entry.js" transform-approved-books-data --data "$json" > "$output_file"
+    configured_books="{}"
```
shouldn't there be tests that exercise the logic of diffing the books against the existing config?

TylerZeroMaster replied:
Thanks for the quick review! Yes, that completely slipped my mind. I added some additional test data for this. The smaller of the new data sets demonstrates what occurs when new versions are present, the other uses some real data to demonstrate that there are no new entries when all ABL versions are already present.

TylerZeroMaster committed:
Update ABL transformation script to read input from files

Use slurp to read input from two files into an array

TylerZeroMaster committed:
Update concourse update-content script

Use process substitution for fetched ABL data

Remove unnecessary check for `detail` property (should be handled by --fail option to curl)

TylerZeroMaster committed:
Update existing ABL test data to include configured books object

Read abl data and configured books from each input file

TylerZeroMaster committed:
Test abl output given abl and configured books as input

TylerZeroMaster committed:
Use some real ABL data for tests

The data was limited to the first 25 entries for brevity

Regardless, the result is the same: no new entries for 20240226.174525
because all the versions from the ABL are already present in configured books

TylerZeroMaster committed:
Add additional input validation to abl transformation

TomWoodward commented:
i think we should wait to merge this until the current batch of updates is done.

@TylerZeroMaster is the new system being used already? are we good to merge this whenever?

TylerZeroMaster commented:
> i think we should wait to merge this until the current batch of updates is done.
> 
> @TylerZeroMaster is the new system being used already? are we good to merge this whenever?

@TomWoodward The new system is in production, but it is not being used yet: there is another PR in progress to update Enki/webhosting pipeline to use the new system as well. We can wait to merge that Enki PR alongside this one.

Additionally, if the ABL on Github is updated, the version on CORGI will need to be updated to match. I have a script that will "fast-forward" the CORGI ABL data to match what is present on Github, but I have been running it manually.

Overall, the changes can be merged anytime, but I would appreciate a little bit of warning ahead of time.

TylerZeroMaster commented on script/transform-approved-books-data.bash:
```diff
@@ -0,0 +1,29 @@
+
+set -e
+
+data="$1"
```
That is a good point. I originally used inputs like this to separate fetching from transformation and the errors that could occur in each process. I updated the transformation script to read input from 2 files and do some additional checks on the input.

TomWoodward commented:
@TylerZeroMaster ok, we will shoot to merge this maybe next week after the current batch of updates is done. 

also the "ABL Schema Tests' check is failing with some jq compile error

TylerZeroMaster committed:
Fix compile error caused by jq version difference

In version 1.7, `if` defaults to returning `.` in the absence of an `else`

In version 1.6, you must explicitly specify `else .`

TylerZeroMaster commented:
> @TylerZeroMaster ok, we will shoot to merge this maybe next week after the current batch of updates is done.
> 
> also the "ABL Schema Tests' check is failing with some jq compile error

Ahh, I think that was because I was using jq version 1.7 which has some changes that must not be backwards compatible. It has been fixed. Let me know when we are ready to merge this change.

staxly[bot] committed:
Merge branch 'main' into corgi-abl

staxly[bot] committed:
Merge branch 'main' into corgi-abl

staxly[bot] committed:
Merge branch 'main' into corgi-abl

staxly[bot] committed:
Merge branch 'main' into corgi-abl

staxly[bot] committed:
Merge branch 'main' into corgi-abl

staxly[bot] committed:
Merge branch 'main' into corgi-abl

staxly[bot] committed:
Merge branch 'main' into corgi-abl

staxly[bot] committed:
Merge branch 'main' into corgi-abl

staxly[bot] committed:
Merge branch 'main' into corgi-abl

staxly[bot] committed:
Merge branch 'main' into corgi-abl

staxly[bot] committed:
Merge branch 'main' into corgi-abl

staxly[bot] committed:
Merge branch 'main' into corgi-abl

staxly[bot] committed:
Merge branch 'main' into corgi-abl

staxly[bot] committed:
Merge branch 'main' into corgi-abl

staxly[bot] committed:
Merge branch 'main' into corgi-abl

staxly[bot] committed:
Merge branch 'main' into corgi-abl

staxly[bot] committed:
Merge branch 'main' into corgi-abl
